recipes/noresm2-lm-hist-vmo/recipe.py

```diff
@@ -0,0 +1,50 @@
+import datetime
+import pandas as pd
+import xarray as xr
+from pangeo_forge_recipes.patterns import ConcatDim, FilePattern
+from pangeo_forge_recipes.recipes.xarray_zarr import XarrayZarrRecipe
+import zarr
+
+BASE_URL = 'http://esgf-data3.ceda.ac.uk/thredds/fileServer/esg_cmip6/CMIP6/CMIP/NCC/NorESM2-LM/historical/r1i1p1f1/Omon/vmo/gr/v20190815/vmo_Omon_NorESM2-LM_historical_r1i1p1f1_gr_'
+
+dates = pd.date_range("1850-01", "2010-01", freq="10YS")
+
+time_concat_dim = ConcatDim("time", keys=dates)
+
+def make_url(time):
+    """With a start date as input, return a url terminating in
+    ``{start}-{end}.nc`` where end is 10 years after the start
+    date for years other than 2010. If the start date is 2010,
+    the end date will be 5 years after the start date.
+    
+    :param date: The start date.
+    """
+    # assign 10 year interval for all years aside from 2010
+    freq = "10YS" if time.year != 2010 else "5YS"
+
+    # make a time range based on the assigned interval
+    start, end = pd.date_range(time, periods=2, freq=freq)
+
+    # subtract one day from the end of the range
+    end = end - datetime.timedelta(days=1)
+
+    # return the url with the timestamp in '%Y%m' format
+    return f"{BASE_URL}{start.strftime('%Y%m')}-{end.strftime('%Y%m')}.nc"
+
+
+pattern = FilePattern(make_url, time_concat_dim)
+
+target_chunks = {'time': 3, 'bnds': 2, 'lev': 70, 'j': 385, 'i': 360, 'vertices': 4}
+
+# the netcdf lists some of the coordinate variables as data variables. This is a fix which we want to apply to each chunk.
+def set_bnds_as_coords(ds):
+    new_coords_vars = [var for var in ds.data_vars if 'bnds' in var or 'bounds' in var]
+    ds = ds.set_coords(new_coords_vars)
+    return ds
+
+recipe = XarrayZarrRecipe(
+    pattern,
+    target_chunks=target_chunks,
+    process_chunk=set_bnds_as_coords,
```

Comment:
Is this our first recipe run with a `process_chunks` function defined in recipe.py? If so that could be the source of the problem.

Reply:
Yes I believe this is the first recipe run with a `process_chunk` callable.

Reply:
Two points:
- we should be able to handle `preocess_chunk`. If we can't, it's a bug we need to fix
- you may be able to bypass this function by adding `xarray_open_kwargs={'decode_coords': 'all'}` to the recipe constructor

Reply:
Alright! Removing `process_chunk` and adding `xarray_open_kwargs={'decode_coords': 'all'}` to the recipe has these errors upon execution:

```python
KeyError: [<class 'h5netcdf.core.File'>, ('/var/folders/p2/czdpkv7x0q3gfchwp2nw54g40000gn/T/tmpogf8fl96/9Emdnr5A/35421a534d9226276baf6ed1c802b2c1-http_esgf-data3.ceda.ac.uk_thredds_fileserver_esg_cmip6_cmip6_cmip_ncc_noresm2-lm_historical_r1i1p1f1_omon_vmo_gr_v20190815_vmo_omon_noresm2-lm_historical_r1i1p1f1_gr_185001-185912.nc',), 'r', (('decode_vlen_strings', True), ('invalid_netcdf', None))]

OSError: Unable to open file (file signature not found)

```

Reply:
Can you open the original file downloaded directly from the server?

Reply:
Yes, and `wget` also works for this file/URL. I've also transferred to the `http://esgf-data3.ceda.ac.uk/` node because `http://noresg.nird.sigma2.no/` keeps having intermittent outages.

Reply:
There error indicates that the file we are caching is not a valid netCDF4 file. Maybe it's a netCDF3 file?

Reply:
The first four bytes of the file are `211   H   D   F` which accdg to [this guide](https://docs.unidata.ucar.edu/netcdf-c/current/faq.html#How-can-I-tell-which-format-a-netCDF-file-uses) corresponds to netCDF4, but this method can't differentiate bet netCDF-4 and netCDF-4 classic, as well as HDF5 files.